vendor/go.etcd.io/etcd/client/v3/kubernetes.go

```diff
@@ -0,0 +1,179 @@
+package clientv3
+
+import (
+	"context"
+
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+)
+
+func NewKubernetes(c *Client) Kubernetes {
+	return &kubernetes{kv: RetryKVClient(c), watcher: c}
+}
+
+type Kubernetes interface {
+	Get(ctx context.Context, key string, revision int64) (kv *mvccpb.KeyValue, rev int64, err error)
+	List(ctx context.Context, prefix string, revision, limit int64, fromKey string) (kvs []*mvccpb.KeyValue, count, rev int64, err error)
+	Count(ctx context.Context, prefix string) (int64, error)
+	OptimisticCreate(ctx context.Context, key string, value []byte) (success bool, rev int64, err error)
+	OptimisticUpdate(ctx context.Context, key string, value []byte, expectedRevision int64) (kv *mvccpb.KeyValue, success bool, rev int64, err error)
+	OptimisticDelete(ctx context.Context, key string, expectedRevision int64) (kv *mvccpb.KeyValue, success bool, rev int64, err error)
+
+	// Consider merging
+	WatchPrefix(ctx context.Context, prefix string, rev int64) WatchChan
+	WatchKey(ctx context.Context, key string, rev int64) WatchChan
+	RequestProgress(ctx context.Context) error
+
+	// Deprecated: Using leases breaks the implicit Kubernetes-etcd contract.
```

Comment:
Is the method deprecated? Or rather maybe instead we want to change that on the etcd side?
The concept of leases themselves we actually need.

Reply:
Method is deprecated because allows K8s to assign multiple keys to one lease. Leases are revoked in single transaction causing multiple deletion to happen in one revision. Kubernetes needs to either:
* Use one lease per key, changed in https://github.com/kubernetes/kubernetes/issues/47532 due to performance problems.
* Change watch protocol, for example utilize bookmarks as it was proposed in https://github.com/kubernetes/kubernetes/issues/110210#issuecomment-1370779174
* Don't use etcd leases, use K8s leases as it was proposed https://github.com/kubernetes/kubernetes/issues/110210#issuecomment-1405635208

Reply:
I've seen all of those, but those are all k8s-side changes.
Why we're implicitly excluding:
- change etcd to not delete all objects of a single lease in a single transation
?

Reply:
Ok, added.

Reply:
I'd be careful with calling something deprecated without having a well-agreed-on alternative for that. Especially, for a feature that is currently in use like Leases.

Reply:
Done.